executor/src/witgen/jit/block_machine_processor.rs

```diff
@@ -0,0 +1,292 @@
+use std::collections::HashSet;
+
+use powdr_ast::analyzed::AlgebraicReference;
+use powdr_number::FieldElement;
+
+use crate::witgen::{machines::MachineParts, FixedData};
+
+use super::{
+    affine_symbolic_expression::Effect,
+    cell::Cell,
+    witgen_inference::{FixedEvaluator, WitgenInference},
+};
+
+struct BlockMachineProcessor<'a, T: FieldElement> {
+    fixed_data: &'a FixedData<'a, T>,
+    machine_parts: MachineParts<'a, T>,
+    block_size: usize,
+    latch_row: usize,
+    inputs: Vec<&'a AlgebraicReference>,
+}
+
+impl<T: FieldElement> BlockMachineProcessor<'_, T> {
+    pub fn generate_code(&self) -> Vec<Effect<T, Cell>> {
+        let known_cells = self
+            .inputs
+            .iter()
+            .map(|input| Cell::from_reference(*input, self.latch_row as i32))
+            .collect::<HashSet<_>>();
+        let ref_eval = FixedEvaluatorForFixedData(self.fixed_data);
+        let mut witgen = WitgenInference::new(self.fixed_data, ref_eval, known_cells);
+
+        let mut complete = HashSet::new();
+        for i in 0.. {
+            let mut progress = false;
+
+            let rows = if i % 2 == 0 {
+                (0..self.block_size).collect::<Vec<_>>()
+            } else {
+                // A common pattern in block machines is that values need to be propagated backwards.
+                // As a performance optimization, we process the rows in reverse order every other iteration.
```

Comment:
Is that important in this code?

Reply:
I thought why not, but sure, let's remove it for now.